RFC.md

```diff
@@ -0,0 +1,466 @@
+        return self.decoder.caches_are_enabled()
+
+	def reset_caches(self):
+        """Reset the key value caches."""
+        self.decoder.reset_caches()
+
+	def forward(
+		self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs, # no need for encoder_mask
+    ) -> Tensor:
+
+    if 	encoder_input is not None:
+		encoder_embeds = self.encoder(**encoder_input)
+
+	tok_mask = tok != self.encoder_token
+	embeds = self.tok_embeddings(tokens[tok_mask])
+	embeds = self._merge_embeds(embeds, encoder_embeds, tok_mask)
+
+	output = self.decoder(tokens, mask, input_pos)
+	return output
+
+ ```
+
+
+### Fusion Layer
+This module's primary use case is for DeepFusion modules where you need to add new layers within the LLM decoder. Instead of adding additional layers directly, which would change layer counts and break checkpoint keys for the text model, FusionLayer fuses the new layer to an existing layer that it modifies. This fused layer is now treated as a single layer. The new parameters are returned with `fusion_params`. Two state dict hooks are used here so that the original layer doesn't have it's name modified in the `state_dict`.
+
+```python
+
+class FusionLayer(nn.Module):
+    """Fusion layer as introduced in `Flamingo: a Visual Language Model for
+    Few-Shot Learning <https://arxiv.org/abs/2204.14198>`_.
+
+    Deep Fusion model architectures combine pretrained encoder models with
+    pretrained language models by infusing the encoder outputs into the middle
+    layers of the LLM. This allow the language model to interpret the encoder
+    outputs as text, to "understand" different modalities that you can train an
+    encoder for. To enable the language model to adapt to the encoder outputs,
+    the FusionLayer inserts a new learnable layer between the decoder (language
+    model) layers to learn to combine the encoder outputs and decoder
+    activations. The fusion layer can be inserted before or after the decoder
+    layer, in Flamingo they insert before.
+
+    Args:
+        layer (nn.Module): original decoder layer
+        fusion_layer (nn.Module): new fusion layer
+        fusion_first (bool): boolean to insert fusion layer before or after
+	        the decoder layer.
+    """
+
+    def __init__(
+        self,
+        layer: nn.Module,
+        fusion_layer: nn.Module,
+        fusion_first: bool = True
+    ):
+        super().__init__()
+        self.layer = layer
+        self.fusion_layer = fusion_layer
+        self.fusion_first = fusion_first
+        self.register_state_dict_post_hook(Test._state_dict_hook)
+        # TODO: Switch to register_load_state_dict_pre_hook in v2.5
+        self._register_load_state_dict_pre_hook(
+	        Test._load_state_dict_hook,
+	        with_module=True
+	    )
+
+    def _state_dict_hook(self, destination, prefix, keep_vars):
```

Comment:
Can you help explain the params?

Reply:
OK now I kinda take back my previous comment. Like if we have to define state dict hooks anyways then we can't really claim that things work out perfectly anyways.. in that case why do we care that the layer numbers line up between the text version and the multimodal version? (We can just remap them in the state dict hook however we want)

Reply:
> Can you help explain the params?

They are the params that pytorch requires for this hook. destination is the state_dict, prefex is any string to prepend to state_dict keys. keep_vars specifies whether tensors in the state_dict should be detached from the computation graph or not.

@ebsmothers 
This is a very small localized change to the state dict that only happens in this module the same way every time. Re-mapping layer numbers affects the entire state dict (instead of a few layers) and will change which layer number you're talking about when debugging if looking at the model vs the state dict. All that is happening here is I'm removing the "layer" variable name that FusionLayer adds to the original layer.